src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/SMDH.java

```diff
@@ -0,0 +1,167 @@
+
+import com.panda3ds.pandroid.data.game.GameRegion;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+public class SMDH {
+    private static final int ICON_SIZE = 48;
+    private static final int META_OFFSET = 0x8 ;
+    private static final int META_REGION_OFFSET = 0x2018;
+    private static final int IMAGE_OFFSET = 0x24C0;
+
+    private int metaIndex = 1;
+    private final ByteBuffer smdh;
+    private final String[] title = new String[12];
+    private final String[] publisher = new String[12];
+    private final int[] icon;
+
+    private final GameRegion region;
+
+    public SMDH(byte[] source){
+        smdh = ByteBuffer.allocate(source.length);
+        smdh.position(0);
+        smdh.put(source);
+        smdh.position(0);
+
+        region = parseRegion();
+        icon = parseIcon();
+        parseMeta();
+    }
+
+    private GameRegion parseRegion(){
+        GameRegion region;
+        smdh.position(META_REGION_OFFSET);
+
+        int regionMasks = smdh.get() & 0xFF;
+
+        final boolean japan = (regionMasks & 0x1) != 0;
+        final boolean northAmerica = (regionMasks & 0x2) != 0;
+        final boolean europe = (regionMasks & 0x4) != 0;
+        final boolean australia = (regionMasks & 0x8) != 0;
+        final boolean china = (regionMasks & 0x10) != 0;
+        final boolean korea = (regionMasks & 0x20) != 0;
+
+        final boolean taiwan = (regionMasks & 0x40) != 0;
+        if (northAmerica) {
+            region = GameRegion.NorthAmerican;
+        } else if (europe) {
+            region = GameRegion.Europe;
+        } else if (australia) {
+            region = GameRegion.Australia;
+        } else if (japan) {
+            region = GameRegion.Japan;
+            metaIndex = 0;
+        } else if (korea) {
+            metaIndex = 7;
+            region = GameRegion.Korean;
+        } else if (china) {
+            metaIndex = 6;
+            region = GameRegion.China;
+        } else if (taiwan) {
+            metaIndex = 6;
+            region = GameRegion.Taiwan;
+        } else {
+            region = GameRegion.None;
+        }
+
+        return region;
+    }
+
+    private void parseMeta(){
+
+        for (int i = 0; i < 12; i++){
+            smdh.position(META_OFFSET + (512*i) + 0x80);
+            byte[] data = new byte[0x100];
+            smdh.get(data);
+            title[i] = convertString(data).replaceAll("\n", " ");
+        }
+
+        for (int i = 0; i < 12; i++){
+            smdh.position(META_OFFSET + (512 * i) + 0x180);
+            byte[] data = new byte[0x80];
+            smdh.get(data);
+            publisher[i] = convertString(data);
+        }
+    }
+
+    private int[] parseIcon() {
+        int[] icon = new int[ICON_SIZE*ICON_SIZE];
+        smdh.position(0);
+
+        for (int x = 0; x < ICON_SIZE; x++) {
+            for (int y = 0; y < ICON_SIZE; y++) {
+                int curseY = y & ~7;
+                int curseX = x & ~7;
```

Comment:
"Curse"?

Reply:
i'll passed 8 hours making this code, and forgets to rename vars

Reply:
i'll rename